packages/common-ts/src/base-service/base-service-v2.ts

```diff
@@ -0,0 +1,232 @@
+      program.addOption(
+        new Option(`--${optionName.toLowerCase()}`, `${optionSpec.desc}`).env(
+          `${params.name
+            .replace(/-/g, '_')
+            .toUpperCase()}__${optionName.toUpperCase()}`
+        )
+      )
+    }
+
+    const longestMetricNameLength = Object.keys(params.metricsSpec).reduce(
+      (acc, key) => {
+        const nameLength = snakeCase(key).length
+        if (nameLength > acc) {
+          return nameLength
+        } else {
+          return acc
+        }
+      },
+      0
+    )
+
+    program.addHelpText(
+      'after',
+      `\nMetrics:\n${Object.entries(params.metricsSpec)
+        .map(([metricName, metricSpec]) => {
+          const parsedName = snakeCase(metricName)
+          return `  ${parsedName}${' '.repeat(
+            longestMetricNameLength - parsedName.length + 2
+          )}${metricSpec.desc} (type: ${metricSpec.type.name})`
+        })
+        .join('\n')}
+      `
+    )
+
+    // Load all configuration values from the environment and argv.
+    program.parse()
+    dotenv.config()
+    const config = new Config(params.name)
+    config.load({
+      env: true,
+      argv: true,
+    })
+
+    // Clean configuration values using the options spec.
+    // Since BCFG turns everything into lower case, we're required to turn all of the input option
+    // names into lower case for the validation step. We'll turn the names back into their original
+    // names when we're done.
+    const cleaned = cleanEnv<TOptions>(
+      { ...config.env, ...config.args },
+      Object.entries(params.optionsSpec || {}).reduce((acc, [key, val]) => {
+        acc[key.toLowerCase()] = val.validator({
+          desc: val.desc,
+          default: val.default,
+        })
+        return acc
+      }, {}) as any,
+      Object.entries(params.options || {}).reduce((acc, [key, val]) => {
+        acc[key.toLowerCase()] = val
+        return acc
+      }, {}) as any
+    )
+
+    // Turn the lowercased option names back into camelCase.
+    this.options = Object.keys(params.optionsSpec || {}).reduce((acc, key) => {
+      acc[key] = cleaned[key.toLowerCase()]
+      return acc
+    }, {}) as TOptions
+
+    // Create the metrics objects.
+    this.metrics = Object.keys(params.metricsSpec || {}).reduce((acc, key) => {
+      const spec = params.metricsSpec[key]
+      acc[key] = new spec.type({
+        name: `${snakeCase(params.name)}_${snakeCase(key)}`,
+        help: spec.desc,
+        labelNames: spec.labels || [],
+      })
+      return acc
+    }, {}) as TMetrics
+
+    this.logger = new Logger({ name: params.name })
+  }
+
+  /**
+   * Runs the main function. If this service is set up to loop, will repeatedly loop around the
+   * main function. Will also catch unhandled errors.
+   */
+  public run(): void {
+    const _run = async () => {
+      if (this.init) {
+        this.logger.info('initializing service')
```

Comment:
Does the logger indicate which service is being initialized? This would be pretty useful info to have.

Reply:
Yeah the logger includes the name of the service in each log line, so it will be clear which service is being initialized.